content/6_Advanced/Vectorization.mdx

```diff
@@ -133,3 +133,265 @@ instead of faster (ex. see
+The bitmask for each query is configured such that each suggested ligature toggles a specific bit.
+
+3. **Counting Occurrences Using SIMD**:
+- The core of the solution utilizes SIMD instructions to parallelize the counting process.
+The program processes the corpus in chunks, accumulating counts for each query simultaneously using AVX2 intrinsics.
+
+**Time Complexity:** $\mathcal{O}(N + Q * K)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#pragma GCC target("avx2")
+#include <immintrin.h>
+#include <iostream>
+#include <string>
+#include <vector>
+using namespace std;
+
+#define rep(i, from, to) for (int i = from; i < (to); ++i)
+#define trav(a, x) for (auto &a : x)
+#define all(x) x.begin(), x.end()
+#define sz(x) (int)(x).size()
+typedef long long ll;
+typedef pair<int, int> pii;
+typedef vector<int> vi;
+
+const int SIMD_BITS = 256;
+const int MAXQ = 100'000;
+
+typedef __m256i mi;
+
+void set_bit(mi &m, size_t ind) {
+	union {
+		mi m;
+		uint32_t ar[SIMD_BITS / 32];
+	} u;
+	u.m = m;
+	u.ar[ind / 32] = (uint32_t)(u.ar[ind / 32] | (1 << (ind % 32)));
+	m = u.m;
+}
+
+void printm(mi m) {
+	union {
+		mi m;
+		uint32_t ar[SIMD_BITS / 32];
+	} u;
+	u.m = m;
+	rep(i, 0, SIMD_BITS) { cout << ((u.ar[i / 32] >> (i % 32)) & 1); }
+	cout << endl;
+}
+
+mi active[MAXQ / SIMD_BITS + 1][26 * 26 + 1]{};  // ~26*26 * MAXQ / 8 bytes
+
+int main() {
+	cin.sync_with_stdio(false);
+	cin.exceptions(cin.failbit);
+	int N, Q, K;
+	cin >> N >> Q >> K;
+	string str, s;
+	cin >> str;
+	N--;
+	int padN = (N + 31) & -32;
+	int qchunks = Q / SIMD_BITS + 1;
+	vector<short> pairs(padN);
+	rep(i, 0, N) {
+		int a = str[i] - 'a';
+		int b = str[i + 1] - 'a';
+		pairs[i] = (short)((26 * a + b) * sizeof(mi));
+	}
+	rep(i, N, padN) { pairs[i] = 26 * 26 * sizeof(mi); }
+
+	rep(i, 0, Q) {
+		cin >> s;
+		rep(j, 0, K) {
+			int a = s[2 * j] - 'a';
+			int b = s[2 * j + 1] - 'a';
+			set_bit(active[i / SIMD_BITS][26 * a + b], i % SIMD_BITS);
+		}
+	}
+
+	vi res(qchunks * SIMD_BITS);
+	rep(k, 0, qchunks) {
+		char *act = (char *)active[k];
+		int i = 0;
+		mi zero = _mm256_setzero_si256();
+		mi acc0 = zero, acc1 = zero, acc2 = zero, acc3 = zero, acc4 = zero;
+		mi accs[30]{}, ts[30]{};
+		mi picked = zero;
+		mi carry;
+#define SET0                                                                   \
+	carry = picked = _mm256_andnot_si256(picked, *(mi *)(act + pairs[i++]))
+#define ADD2(t, acc)                                                           \
+	{                                                                          \
+		mi x = t ^ carry, y = (x & acc) | (t & carry);                         \
+		acc = x ^ acc, carry = y;                                              \
+	}  // (carry, acc) = (t + acc + carry)
+#define ADD(ind) ADD2(t##ind, acc##ind)
+		while (i < padN) {
+			mi t0, t1, t2, t3, t4;
+			SET0;
+			t0 = carry;
+			SET0;
+			ADD(0);
+			t1 = carry;
+			SET0;
```

Comment:
what the heck

Reply:
in reference to this: https://en.wikipedia.org/wiki/Streaming_SIMD_Extensions from the module?

Reply:
Sorry, this was the solution that the content author linked to.I apologize for not commenting beforehand. 
Here is my understanding:
It prepares the data structures (pairs array, active array) and initializes N, Q, and K for processing ligature sets using SIMD instructions. It uses padding for alignment, vectorized operations, and direct memory manipulation for performance optimizations. This setup is important for the following SIMD-based computations to efficiently count ligature occurrences in the given corpus and ligature sets.